File: src/System/Mesos/Raw/TrafficControlStatistics.hs
module System.Mesos.Raw.TrafficControlStatistics where
import           System.Mesos.Internal

type TrafficControlStatisticsPtr = Ptr TrafficControlStatistics

foreign import ccall unsafe "ext/types.h toTrafficControlStatistics" c_toTrafficControlStatistics
  :: Ptr CChar -- ^ id
  -> CInt -- ^ idlen
  -> Ptr CULong -- ^ backlog
  -> Ptr CULong -- ^ bytes
  -> Ptr CULong -- ^ drops
  -> Ptr CULong -- ^ overlimits
  -> Ptr CULong -- ^ packets
  -> Ptr CULong -- ^ qlen
  -> Ptr CULong -- ^ ratebps
  -> Ptr CULong -- ^ ratepps
  -> Ptr CULong -- ^ requeues
  -> IO TrafficControlStatisticsPtr

foreign import ccall unsafe "ext/types.h fromTrafficControlStatistics" c_fromTrafficControlStatistics
  :: TrafficControlStatisticsPtr
  -> Ptr (Ptr CChar) -- ^ id
  -> Ptr CInt -- ^ idlen
  -> Ptr CULong -- ^ backlog
  -> Ptr CBool  -- ^ backlogSet
  -> Ptr CULong -- ^ bytes
  -> Ptr CBool -- ^ bytesSet
  -> Ptr CULong -- ^ drops
  -> Ptr CBool -- ^ bytesSet
  -> Ptr CULong -- ^ overlimits
  -> Ptr CBool -- ^ overlimitsSet
  -> Ptr CULong -- ^ packets
  -> Ptr CBool -- ^ packetsSet
  -> Ptr CULong -- ^ qlen
  -> Ptr CBool -- ^ qlenSet
  -> Ptr CULong -- ^ ratebps
  -> Ptr CBool -- ^ ratebpsSet
  -> Ptr CULong -- ^ ratepps
  -> Ptr CBool -- ^ rateppsSet
  -> Ptr CULong -- ^ requeues
  -> Ptr CBool -- ^ requeuesSet
  -> IO ()

foreign import ccall unsafe "ext/types.h destroyTrafficControlStatistics" c_destroyTrafficControlStatistics
  :: TrafficControlStatisticsPtr
  -> IO ()

instance CPPValue TrafficControlStatistics where
  marshal tcs = do
    (idp, idl) <- cstring $ trafficControlStatisticsIDId' tcs
    blgp <- maybeCULong trafficControlStatisticsBacklog
    bytesp <- maybeCULong trafficControlStatisticsBytes
    dropsp <- maybeCULong trafficControlStatisticsDrops
    overlimitsp <- maybeCULong trafficControlStatisticsOverlimits
    packetsp <- maybeCULong trafficControlStatisticsPackets
    qlenp <- maybeCULong trafficControlStatisticsQlen
    ratebpsp <- maybeCULong trafficControlStatisticsRatebps
    rateppsp <- maybeCULong trafficControlStatisticsRatepps
    requeuesp <- maybeCULong trafficControlStatisticsRequeues

    liftIO $ c_toTrafficControlStatistics idp (fromIntegral idl)
               blgp
               bytesp
               dropsp
               overlimitsp
               packetsp
               qlenp
               ratebpsp
               rateppsp
               requeuesp
   where
     maybeCULong f = allocMaybe . fmap CULong $ f tcs

  unmarshal tcs = do
    (idpp, idlp) <- arrayPair
    backlogp <- alloc
    backlogSetp <- alloc
    bytesp <- alloc
    bytesSetp <- alloc
    dropsp <- alloc
    dropsSetp <- alloc
    overlimitsp <- alloc
    overlimitsSetp <- alloc
    packetsp <- alloc
    packetsSetp <- alloc
    qlenp <- alloc
    qlenSetp <- alloc
    ratebpsp <- alloc
    ratebpsSetp <- alloc
    rateppsp <- alloc
    rateppsSetp <- alloc
    requeuesp <- alloc
    requeuesSetp <- alloc
    liftIO $ c_fromTrafficControlStatistics tcs
               idpp idlp
               backlogp backlogSetp
               bytesp bytesSetp
               dropsp dropsSetp
               overlimitsp overlimitsSetp
               packetsp packetsSetp
               qlenp qlenSetp
               ratebpsp ratebpsSetp
               rateppsp rateppsSetp
               requeuesp requeuesSetp

    tid <- peekCString (idpp, idlp)
    backlog <- toWord64 <$> peekMaybePrim backlogp backlogSetp
    bytes <- toWord64 <$> peekMaybePrim bytesp bytesSetp
    drops <- toWord64 <$> peekMaybePrim dropsp dropsSetp
    overlimits <- toWord64 <$> peekMaybePrim overlimitsp overlimitsSetp
    packets <- toWord64 <$> peekMaybePrim packetsp packetsSetp
    qlen <- toWord64 <$> peekMaybePrim qlenp qlenSetp
    ratebps <- toWord64 <$> peekMaybePrim ratebpsp ratebpsSetp
    ratepps <- toWord64 <$> peekMaybePrim rateppsp rateppsSetp
    requeues <- toWord64 <$> peekMaybePrim requeuesp requeuesSetp
    return $ TrafficControlStatistics tid backlog bytes drops overlimits packets
               qlen ratebps ratepps requeues
   where
    toWord64 mx = case mx of
      Nothing -> Nothing
      Just (CULong x) -> Just x

  destroy = c_destroyTrafficControlStatistics

File: ext/types/traffic_control_statistics.cpp
#include <iostream>
#include "types.h"

using namespace mesos;

TrafficControlStatisticsPtr toTrafficControlStatistics(char* id,
                                                       int idLen,
                                                       unsigned long* backlog,
                                                       unsigned long* bytes,
                                                       unsigned long* drops,
                                                       unsigned long* overlimits,
                                                       unsigned long* packets,
                                                       unsigned long* qlen,
                                                       unsigned long* ratebps,
                                                       unsigned long* ratepps,
                                                       unsigned long* requeues)
{
  TrafficControlStatisticsPtr tcs = new TrafficControlStatistics();

  tcs->set_id(id, idLen);

  if (backlog != NULL)
    tcs->set_backlog(*backlog);

  if (bytes != NULL)
    tcs->set_bytes(*bytes);

  if (drops != NULL)
    tcs->set_drops(*drops);

  if (overlimits != NULL)
    tcs->set_overlimits(*overlimits);

  if (packets != NULL)
    tcs->set_packets(*packets);

  if (qlen != NULL)
    tcs->set_qlen(*qlen);

  if (ratebps != NULL)
    tcs->set_ratebps(*ratebps);

  if (ratepps != NULL)
    tcs->set_ratepps(*ratepps);

  if (requeues != NULL)
    tcs->set_requeues(*requeues);

  return tcs;
}

void fromTrafficControlStatistics(TrafficControlStatisticsPtr tcs,
                                  char** id,
                                  int* idLen,
                                  unsigned long* backlog,
                                  bool* backlogSet,
                                  unsigned long* bytes,
                                  bool* bytesSet,
                                  unsigned long* drops,
                                  bool* dropsSet,
                                  unsigned long* overlimits,
                                  bool* overlimitsSet,
                                  unsigned long* packets,
                                  bool* packetsSet,
                                  unsigned long* qlen,
                                  bool* qlenSet,
                                  unsigned long* ratebps,
                                  bool* ratebpsSet,
                                  unsigned long* ratepps,
                                  bool* rateppsSet,
                                  unsigned long* requeues,
                                  bool* requeuesSet)
{
  std::string *i = tcs->mutable_id();
  *id = (char*) i->data();
  *idLen = i->size();

  *backlogSet = false;
  *bytesSet = false;
  *dropsSet = false;
  *overlimitsSet = false;
  *packetsSet = false;
  *qlenSet = false;
  *ratebpsSet = false;
  *rateppsSet = false;
  *requeuesSet = false;

  if (tcs->has_backlog())
    {
      *backlog = tcs->backlog();
      *backlogSet = true;
    }
  if (tcs->has_bytes())
    {
      *bytes = tcs->bytes();
      *bytesSet = true;
    }

  if (tcs->has_drops())
    {
      *drops = tcs->drops();
      *dropsSet = true;
    }

  if (tcs->has_overlimits())
    {
      *overlimits = tcs->overlimits();
      *overlimitsSet = true;
    }

  if (tcs->has_packets())
    {
      *packets = tcs->packets();
      *packetsSet = true;
    }

  if (tcs->has_qlen())
    {
      *qlen = tcs->qlen();
      *qlenSet = true;
    }

  if (tcs->has_ratebps())
    {
      *ratebps = tcs->ratebps();
      *ratebpsSet = true;
    }

  if (tcs->has_ratepps())
    {
      *ratepps = tcs->ratepps();
      *rateppsSet = true;
    }

  if (tcs->has_requeues())
    {
      *requeues = tcs->requeues();
      *requeuesSet = true;
    }
}

void destroyTrafficControlStatistics(TrafficControlStatisticsPtr tcs)
{
  delete tcs;
}
